### PR TITLE
fix: support discriminated dynamic models

### PR DESCRIPTION
**Pull request by jorgerangel-msft**
This PR fixes some edge cases when generating discriminated dynamic models, fixes an issue where the `GetUtf8Bytes` helper method wasn't being generated in cases where a non-dynamic model contained model properties, and adds test cases for dynamic models that publicly support additional properties.

fixes: https://github.com/microsoft/typespec/issues/8477, https://github.com/microsoft/typespec/issues/8470

### CODE REVIEW

File: packages/http-client-csharp/generator/Microsoft.TypeSpec.Generator.ClientModel/src/Providers/ModelSerializationExtensionsDefinition.cs [hunk 1]
```diff
@@ -157,6 +157,7 @@ protected override MethodProvider[] BuildMethods()
                 BuildWriteNumberValueMethodProvider(),
                 BuildWriteObjectValueMethodGeneric(),
                 BuildWriteObjectValueMethodProvider(),
+                BuildGetUtf8BytesMethodProvider(),
```
Comment (JoshLove-msft) [inline packages/http-client-csharp/generator/Microsoft.TypeSpec.Generator.ClientModel/src/Providers/ModelSerializationExtensionsDefinition.cs]: Wonder if we there is an issue in the logic of HasDynamicModels. Maybe we should just set the property to true in MarkModelsAsDynamicRecursive?

File: packages/http-client-csharp/generator/Microsoft.TypeSpec.Generator.ClientModel/src/Providers/ScmModelProvider.cs [hunk 2]
```diff
@@ -253,6 +261,25 @@ private void UpdateFullConstructorParameters()
                 }
             }
 
+            // Update the initializer to include the json patch parameter as an argument
+            if (FullConstructor.Signature.Initializer != null)
+            {
+                var updatedArguments = new List<ValueExpression>(FullConstructor.Signature.Initializer.Arguments.Count);
+                foreach (var argument in FullConstructor.Signature.Initializer.Arguments)
+                {
+                    if (argument is VariableExpression { Declaration.RequestedName: AdditionalBinaryDataParameterName })
```
Comment (JoshLove-msft) [inline packages/http-client-csharp/generator/Microsoft.TypeSpec.Generator.ClientModel/src/Providers/ScmModelProvider.cs]: Is the argument at all connected to the underlying RawDataField? If so, could we use that instead?

**Review by JoshLove-msft — APPROVED** [review]

File: packages/http-client-csharp/generator/Microsoft.TypeSpec.Generator.ClientModel/src/Providers/ModelSerializationExtensionsDefinition.cs (hunk 1 above)
Comment (jorgerangel-msft) [reply packages/http-client-csharp/generator/Microsoft.TypeSpec.Generator.ClientModel/src/Providers/ModelSerializationExtensionsDefinition.cs]: This was actually a bug I found when running the spector scenarios. We are unconditionally adding the BinaryData param to the deserialization method for a model, but conditionally adding this helper method. The other possible fix I thought of is to conditionally add the BinaryData param to the deserialization method only for dynamic models.

Comment (JoshLove-msft) [reply packages/http-client-csharp/generator/Microsoft.TypeSpec.Generator.ClientModel/src/Providers/ModelSerializationExtensionsDefinition.cs]: Ah ok, yeah I added the BinaryData param unconditionally for the sake of simplicity but hadn't considered the nested scenario where we actually need to convert back to Utf8Bytes. I would just remove it for non-dynamic models.

File: packages/http-client-csharp/generator/Microsoft.TypeSpec.Generator.ClientModel/src/Providers/MrwSerializationTypeDefinition.cs [hunk 3]
```diff
@@ -2330,5 +2345,17 @@ private static string BuildJsonPathForElement(string propertySerializedName, Lis
 
             return result;
         }
+
+        private static ValueExpression GetDeserializationMethodInvocationForType(
```
Comment (JoshLove-msft) [inline packages/http-client-csharp/generator/Microsoft.TypeSpec.Generator.ClientModel/src/Providers/MrwSerializationTypeDefinition.cs:2349]: Hmm does this need to be internal if we are using it in ScmMethodProviderCollection?

Comment (jorgerangel-msft) [reply packages/http-client-csharp/generator/Microsoft.TypeSpec.Generator.ClientModel/src/Providers/MrwSerializationTypeDefinition.cs:2349]: I have 2 overloads, one that I made internal since I needed similar functionality in ScmMethodProviderCollection. It seems to not complain if I keep the other overload private? Though I can also make it internal for consistency

Comment (JoshLove-msft) [reply packages/http-client-csharp/generator/Microsoft.TypeSpec.Generator.ClientModel/src/Providers/MrwSerializationTypeDefinition.cs:2349]: Oh I hadn't realized there was another overload.
